### PR TITLE
improve unstack error messages

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/DataFrames.jl/issues/3339

### CODE REVIEW

File: src/abstractdataframe/reshape.jl [hunk 1]
```diff
@@ -545,7 +545,8 @@ function _unstack(df::AbstractDataFrame, rowkeys::AbstractVector{Int},
     colref_map = df[col_group_row_idxs, colkey]
     if any(ismissing, colref_map) && !allowmissing
         throw(ArgumentError("Missing value in variable :$(_names(df)[colkey]). " *
-                            "Pass `allowmissing=true` to skip missings."))
+                            "Pass `allowmissing=true` to create a column referring. " *
+                            "to `missing` value"))
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl:549]: Maybe this is more explicit?
```suggestion
                            "Pass `allowmissing=true` to create a :missing column " *
                            "referring to `missing` values."))
```

File: src/abstractdataframe/reshape.jl [hunk 2]
```diff
@@ -579,7 +580,13 @@ function _unstack(df::AbstractDataFrame, rowkeys::AbstractVector{Int},
     # and names between df1 and df2 must be unique
     # here df1 gets proper column-level metadata with :note style
     df1 = df[row_group_row_idxs, g_rowkey.cols]
-    df2 = DataFrame(unstacked_val, Symbol[Symbol(renamecols(x)) for x in colref_map],
+    new_col_names = Symbol[Symbol(renamecols(x)) for x in colref_map]
+    if !allunique(new_col_names)
+        throw(ArgumentError("Non-unique column names produced. " *
+                            "Non equal values in `colkey` were mapped " *
+                            "to the same column name."))
+    end
```
Comment (nalimilan) [inline src/abstractdataframe/reshape.jl:588]: In theory this is breaking. Can we be reasonably sure that there are no legitimate use cases?

Mixing `"a"` and `'a'` or `missing` and `"missing"` in the same column and expecting them to be treated the same seems relatively unlikely, but are there other possible cases?

Comment (bkamins) [reply src/abstractdataframe/reshape.jl:588]: > In theory this is breaking

It is not breaking. This PR only adds more informative error messages. Currently, we have for example (other cases produce similar errors - which are quite confusing - this "confusing" nature of the error messages is the point of why this PR is made as we do not allow passing `makeunique=true`):

```
julia> df = DataFrame(x=1:3, y=["a", "b", "c"], z=11:13)
3×3 DataFrame
 Row │ x      y       z
     │ Int64  String  Int64
─────┼──────────────────────
   1 │     1  a          11
   2 │     2  b          12
   3 │     3  c          13

julia> unstack(df, :x, :y, :z, renamecols=x->"v")
ERROR: ArgumentError: Duplicate variable names: :v. Pass makeunique=true to make them unique using a suffix automatically.
```

> are there other possible cases?

The most likely cases are:
* when a user passes a column renaming function that produces duplicates (e.g. strips some part of the name and user thinks that it is still unique while it is not)
* when column name conflicts with key name.

Comment (bkamins) [reply src/abstractdataframe/reshape.jl:588]: To expand on it. If `unstack` sees `'a'` and `"a"` it puts them in different groups (the behavior is just like `groupby` behaves) and performs unstacking. The error is produced much later when column names are generated for the final data frame that is ultimately returned.

**Review by nalimilan — APPROVED** [review]

File: src/abstractdataframe/reshape.jl [hunk 3]
```diff
@@ -588,6 +595,12 @@ function _unstack(df::AbstractDataFrame, rowkeys::AbstractVector{Int},
         df2 = df2[!, sortperm(col_group_row_idxs)]
     end
 
+    if !isempty(intersect(_names(df1), _names(df2)))
+        throw(ArgumentError("Non-unique column names produced. " *
+                            "Column names created using the `colkey`" *
```
Comment (bkamins) [inline src/abstractdataframe/reshape.jl:600]:
```suggestion
                            "Column names created using the `colkey` " *
```

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**